crates/nu-plugin/src/plugin/persistent.rs

```diff
@@ -0,0 +1,186 @@
+
+        // We need the current environment variables for `python` based plugins
+        // Or we'll likely have a problem when a plugin is implemented in a virtual Python environment.
+        plugin_cmd.envs(envs);
+
+        let program_name = plugin_cmd.get_program().to_os_string().into_string();
+
+        // Run the plugin command
+        let child = plugin_cmd.spawn().map_err(|err| {
+            let error_msg = match err.kind() {
+                std::io::ErrorKind::NotFound => match program_name {
+                    Ok(prog_name) => {
+                        format!("Can't find {prog_name}, please make sure that {prog_name} is in PATH.")
+                    }
+                    _ => {
+                        format!("Error spawning child process: {err}")
+                    }
+                },
+                _ => {
+                    format!("Error spawning child process: {err}")
+                }
+            };
+            ShellError::PluginFailedToLoad { msg: error_msg }
+        })?;
+
+        // Start the plugin garbage collector
+        let gc_config =
+            self.gc_config
+                .lock()
+                .map(|c| c.clone())
+                .map_err(|_| ShellError::NushellFailed {
+                    msg: "plugin gc mutex poisoned".into(),
+                })?;
+        let gc = PluginGc::new(gc_config, &self)?;
+
+        let pid = child.id();
+        let interface =
+            make_plugin_interface(child, Arc::new(PluginSource::new(&self)), Some(gc.clone()))?;
+
+        Ok(RunningPlugin { pid, interface, gc })
+    }
+}
+
+impl RegisteredPlugin for PersistentPlugin {
+    fn identity(&self) -> &PluginIdentity {
+        &self.identity
+    }
+
+    fn is_running(&self) -> bool {
+        // If the lock is poisoned, we return false here. That may not be correct, but this is a
+        // failure state anyway that would be noticed at some point
+        self.running.lock().map(|r| r.is_some()).unwrap_or(false)
+    }
+
+    fn pid(&self) -> Option<u32> {
+        // Again, we return None for a poisoned lock.
+        self.running
+            .lock()
+            .ok()
+            .and_then(|r| r.as_ref().map(|r| r.pid))
+    }
+
+    fn stop(&self) -> Result<(), ShellError> {
+        let mut running = self.running.lock().map_err(|_| ShellError::NushellFailed {
+            msg: format!(
+                "plugin `{}` running mutex poisoned, probably panic during spawn",
+                self.identity.name()
+            ),
+        })?;
+
+        // If the plugin is running, stop its GC, so that the GC doesn't accidentally try to stop
+        // a future plugin
+        if let Some(running) = running.as_ref() {
+            running.gc.stop_tracking();
+        }
+
+        // We don't try to kill the process or anything, we just drop the RunningPlugin. It should
+        // exit soon after
+        *running = None;
```

Comment:
Sorry, can you elaborate more?  After reading 2nd point in your comment, I still failed to understand how it works on plugin side.  Is the logic handled inside `serve_plugin` function?

Reply:
Ok, I get the answer.
When set `*running = None`, then:
```
1. drop runner value 
2. PluginInterface is dropped
3. then PluginInterfaceState is dropped
4. then send goodbye to plugin
```